interfaces/cython/cantera/utils.pyx

```diff
@@ -72,3 +72,78 @@ class CanteraError(RuntimeError):
     pass
 
 cdef public PyObject* pyCanteraError = <PyObject*>CanteraError
+
+cdef anyvalueToPython(string name, CxxAnyValue& v):
+    cdef CxxAnyMap a
+    cdef CxxAnyValue b
+    if v.isScalar():
+        if v.isType[string]():
+            return pystr(v.asType[string]())
+        elif v.isType[double]():
+            return v.asType[double]()
+        elif v.isType[long]():
+            return v.asType[long]()
+        elif v.isType[cbool]():
+            return v.asType[cbool]()
+    elif v.isType[CxxAnyMap]():
+        return anymapToPython(v.asType[CxxAnyMap]())
+    elif v.isType[vector[CxxAnyMap]]():
+        return [anymapToPython(a) for a in v.asType[vector[CxxAnyMap]]()]
+    elif v.isType[vector[double]]():
+        return v.asType[vector[double]]()
+    elif v.isType[vector[string]]():
+        return [pystr(s) for s in v.asType[vector[string]]()]
+    elif v.isType[vector[long]]():
+        return v.asType[vector[long]]()
+    elif v.isType[vector[cbool]]():
+        return v.asType[vector[cbool]]()
+    elif v.isType[vector[CxxAnyValue]]():
+        return [anyvalueToPython(name, b)
+                for b in v.asType[vector[CxxAnyValue]]()]
+    elif v.isType[vector[vector[double]]]():
+        return v.asType[vector[vector[double]]]()
+    elif v.isType[vector[vector[string]]]():
+        return [[pystr(s) for s in row]
+                for row in v.asType[vector[vector[string]]]()]
+    elif v.isType[vector[vector[long]]]():
+        return v.asType[vector[vector[long]]]()
+    elif v.isType[vector[vector[cbool]]]():
+        return v.asType[vector[vector[cbool]]]()
+    else:
+        raise TypeError("Unable to convert value with key '{}' "
+                        "from AnyValue of held type '{}'".format(
+                            pystr(name), v.type_str()))
+
+
+cdef anymapToPython(CxxAnyMap& m):
+    py_items = []
+    m.applyUnits()
+    for item in m:
+        py_items.append((item.second.order(),
+                         pystr(item.first),
+                         anyvalueToPython(item.first, item.second)))
+    py_items.sort()
+    return {key: value for (_, key, value) in py_items}
+
+
+cdef mergeAnyMap(CxxAnyMap& primary, CxxAnyMap& extra):
```

Comment:
Would it make sense to handle that within the C++ layer? Also, as a comment, the `anymapToPython` (or `anymap_to_dict`) applies units whereas `mergeAnyMap` does not.

Regarding my other comment about `getParameters`, a C++ implementation of `merge` would go in hand with having `unique_ptr<AnyMap> parameter() const` returned throughout, which then would mean that you can `AnyMap::merge(const AnyMap& other)` (or similar).

I assume that there may a small performance penalty for smart pointers, but it would be easier to track where objects are initially created (while this may be subjective, it is often difficult to track how things are assembled when the receiving function creates the object which is subsequently added to as in your `getParameters` approach).

Reply:
while my initial comment is mostly moot, I believe this function is mainly used to merge output from `getParameters` and `input`. I believe that it would be nice to merge this by passing an optional flag to `getParameters` where it applies (and handle the merge in C++). This may eliminate the need for this specific Cython function ...

Reply:
The initial reason for this came from the desire to control the order of the output keys. I want the user-defined keys in `input` to generally come last, and the ordering mechanism mostly relies on preserving insertion order. If the base `Reaction` class were responsible for adding in the user-defined keys, then they would come before any keys added by child classes.

I actually ended up finding a hybrid solution that both gets rid of the need to merge the fields defining object with the extra input data at the point of use, and avoids the requirement for the user-accessible function to take an `AnyMap` as it's input, by making the virtual `getParameters` method `protected`, and introducing a non-virtual method with the signature `AnyMap parameters(bool withInput=true)` on the base classes only. This way, the `parameters` method can initialize the `AnyMap`, use the virtual `getParameters` method to populate all of the relevant fields, and then add the extra keys last before returning it.